version/version.go

```diff
@@ -5,98 +5,162 @@
+	// output.
+	longStamp string
+
+	// shortStamp is the short version identifier of the build. If set, it
+	// is returned verbatim by Short() and other functions that return Short()'s
+	// output.
+	shortStamp string
+
+	// gitCommitStamp is the git commit of the github.com/tailscale/tailscale
+	// repository at which Tailscale was built. Its format is the one returned
+	// by `git rev-parse <commit>`. If set, it is used instead of any git commit
+	// information embedded by the Go tool.
+	gitCommitStamp string
+
+	// gitDirtyStamp is whether the git checkout from which the code was built
+	// was dirty. Its value is ORed with the dirty bit embedded by the Go tool.
+	//
+	// We need this because when we build binaries from another repo that
+	// imports tailscale.com, the Go tool doesn't stamp any dirtiness info into
+	// the binary. Instead, we have to inject the dirty bit ourselves here.
+	gitDirtyStamp bool
+
+	// extraGitCommit, is the git commit of a "supplemental" repository at which
+	// Tailscale was built. Its format is the same as gitCommit.
+	//
+	// extraGitCommit is used to track the source revision when the main
+	// Tailscale repository is integrated into and built from another repository
+	// (for example, Tailscale's proprietary code, or the Android OSS
+	// repository). Together, gitCommit and extraGitCommit exactly describe what
+	// repositories and commits were used in a build.
+	extraGitCommitStamp string
+)
+
+var long lazy.SyncValue[string]
+
+// Long returns a full version number for this build, of one of the forms:
+//
+//   - "x.y.z-commithash-otherhash" for release builds distributed by Tailscale
+//   - "x.y.z-commithash" for release builds built with build_dist.sh
+//   - "x.y.z-changecount-commithash-otherhash" for untagged release branch
+//     builds by Tailscale (these are not distributed).
+//   - "x.y.z-changecount-commithash" for untagged release branch builds
+//     built with build_dist.sh
+//   - "x.y.z-devYYYYMMDD-commithash{,-dirty}" for builds made with plain "go
+//     build" or "go install"
+//   - "x.y.z-ERR-BuildInfo" for builds made by plain "go build" or "go install"
+//     that do not embed build metadata
+func Long() string {
+	return long.Get(func() string {
+		if longStamp != "" {
+			return longStamp
+		}
+		bi := getEmbeddedInfo()
+		if !bi.valid {
+			return strings.TrimSpace(tailscaleroot.VersionDotTxt) + "-ERR-BuildInfo"
+		}
+		return fmt.Sprintf("%s-dev%s-t%s%s", strings.TrimSpace(tailscaleroot.VersionDotTxt), bi.commitDate, bi.commitAbbrev(), dirtyString())
+	})
+}
+
+var short lazy.SyncValue[string]
+
+// Short returns a short version number for this build, of the forms:
+//
+//   - "x.y.z" for builds distributed by Tailscale or built with build_dist.sh
+//   - "x.y.z-devYYYYMMDD" for builds made with plain "go build" or "go install"
+//   - "x.y.z-ERR-BuildInfo" for builds made by plain "go build" or "go install"
+//     that do not embed build metadata
+func Short() string {
```

Comment:
out of curiosity, does this allocate? can you add a testing.AllocsPerRun(1000, ...) for this to see?

Reply:
Added test, doesn't allocate.